### ISSUE TITLE
Missing reference cuando usás una lista

### ISSUE THREAD

**Issue by fdodino** [uqbar-project/wollok-ts#165]
La siguiente definición está tirando un error falopa en object:
![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/cda6ad78-b6cf-488c-ba2b-c9e383d0e877)

```wollok
object pepita {
    var energia = 50
    const lista = [1, 2, 3]

    method volar(minutos) {
        energia = energia - minutos * 2
    }

    method comer(gramos) {
        energia = energia + gramos * 5
    }

}
```

el problema es por `const lista = [1, 2, 3]`, hay una validación que no estoy muy seguro qué issue resolvía @PalumboN 

```ts
// TODO: Test if the reference points to the right kind of node
export const missingReference = error<Reference<Node>>(node => !!node.target)
```

**Comment by fdodino** [comment]
Upa, el diagrama dinámico me muestra ésto, hay que ver si es porque se mambea el diagrama o bien porque es un error del intérprete:

![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/563d7ce5-3b97-40dc-91c4-ff45173ed4a4)

**Comment by fdodino** [comment]
Más data de lo raro que se ve: al meter una clase me sigue tirando un missing reference en el object, pero el diagrama dinámico se ve bien

![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/9d97d3ac-b27b-4f63-a1a8-21311526cbfa)

si hago `const p = new Ave()`
![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/a76b52a6-237a-4560-b8c3-86407dfeee38)

**Comment by fdodino** [comment]
La referencia está (en algún momento pensé que no) porque yo puedo evaluar el size, 
![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/1aa0c284-a0b4-4e37-803e-6be2852b3b97)

**Comment by nscarcella** [comment]
Rari.

La validación está porque toda referencia tiene que quedar apuntando a su target, sino quiere decir que algo salió mal en el linkeo.

Pero sin la referencia no deberías poder evaluar el size...

Habría que revisar si, efectivamente, el nodo pepita tiene una referencia targeteando a la lista.

Ayudaría si el error mostrara cuál es la referencia que no tiene target

**Comment by PalumboN** [comment]
Holisssss,

Ese `missingReference` dice que está en la posición `1:1`, lo cual me hace dudar.

Para mí evaluaste algún código que hizo que se rompiera el parser (y no se pueda recuperar) y te quedó choteado el LSP. Cuando pasa eso hay que mirar la consola del server a ver si está tirando error o qué onda.

Seguro es un problema del LSP, en runtime anda todo eso (quiero creer jajaja).

**Comment by nscarcella** [comment]
Si, pero igual eso no tiene sentido porque en 1:1 no hay ninguna
referencia. Hay algo choteado en el reporte

**Comment by PalumboN** [comment]
Sí, eso puede ser porque el error lo tiene el nodo file, o el error vino sin sourceMap y creo que se le ponía `1:1` por default.
Habría que ponerle un enter antes del `object pepita` para checkear si está atado al object o al file.

Eso es algo que estuvimos arreglando con Ivo en el último tiempo, no recuerdo bien cuál es el estado actual.

**Comment by PalumboN** [comment]
@fdodino ese LSP lo levantaste apuntando a la versión `master` de TS?

**Comment by fdodino** [comment]
sí, @PalumboN , es contra master (los últimos cambios que le metí yo)

y efectivamente es en la primera línea del archivo donde se rompe

![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/ce93bba0-cfba-43e5-a5c9-cbb5f8fad031)

**Comment by PalumboN** [comment]
Y en la consola del server no salta nada raro? (del primer VSCode desde donde se levanta el de la extensión)

<img width="1116" alt="image" src="https://github.com/uqbar-project/wollok-ts/assets/4098184/9565b674-b197-4660-87f7-af148aeadf46">

**Comment by fdodino** [comment]
Yo le puse un console.info en el validator:

```
***************************** undefined 29833c2c-a22c-417b-95f9-912f1681d5f0 Reference [Reference]{81d5f0} at home/dodain/workspace/wollok/wollok-workspace/example.wlk:--
```

```ts
export const missingReference = error<Reference<Node>>(node => {
  if (!node.target) console.info('***************************** undefined', node.id, node.kind, node.label)
  return !!node.target
})
```

**Comment by fdodino** [comment]
por las dudas:
![image](https://github.com/uqbar-project/wollok-ts/assets/4549002/0c1e281f-c88b-41c7-8d4f-eb3061a15d27)

**Comment by PalumboN** [comment]
Y cuál es la referencia? Tiene `name`?

**Comment by nscarcella** [comment]
Podrías también imprimir el label del parent de la reference

**Comment by fdodino** [comment]
ahí stringifié todo el node para que lo vean:

```
***  {"metadata":[],"name":"wollok.lang.List","id":"8d76ba4c-8b98-4f9e-b1f5-9935fa813372","scope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[{"containerScope":{"containerScope":{"containerScope":{"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]}]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]}}
Performance metrics
=========================
o- build environment for file 139
o- validation time 0

***  {"metadata":[],"name":"wollok.lang.List","id":"8d901637-1b64-4ca1-b527-18ecffa5e98e","scope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"containerScope":{"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[{"containerScope":{"containerScope":{"containerScope":{"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]}]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]},"contributions":{},"includedScopes":[]}}
Performance metrics
=========================
o- build environment for file 104
o- validation time 0

```

**Comment by fdodino** [comment]
Y el label del parent de la reference no dice mucho, es `***  "[Literal]{1a0ce4} at home/dodain/workspace/wollok/wollok-workspace/example.wlk:11"`. Me rajo porque mi cocina tiene un caño roto...

**Comment by PalumboN** [comment]
Wooo es la referencia a la clase List que referencia el literal.... Raro que no encuentre una clase de Lang

**Comment by nscarcella** [comment]
Uhmmmm... Que falopa.

Podés pasar el scope del literal? El root debería tener List incluído.
Aunque ahora no me acuerdo bien exactamente a qué apuntaba el literal. O
cuándo se linkeaba. Creeeo que a una tupla de clase y elementos.

Dame un rato y miro


N.

**Comment by nscarcella** [comment]
Ok, ahí miré en github (no tengo la compu a mano).

Probablemente la parte relevante esté acá: La linea 102 del Linker:

node.is(Reference) && parent!.is(ParameterizedType)
? parent?.parent.scope
: parent?.scope
),

La cosa es así: La referencia a List tiene que ser un children del Literal
(95% seguro que eso funciona). Cuando se están asignando los scopes, el
scope de la referencia tendría que tener como padre el scope del literal
que eventualmente tiene que tener al root como padre, peeeero capaz al
momento de correr este código todavía no se asignó el scope del padre al
padre.

Entonces yo miraría si:
A) El scope de la referencia eventualmente llega al root (que es el que
tiene que tener definido "wollok" adentro) y
B) Si eso no es así, habría que ver qué pasó...

**Comment by PalumboN** [comment]
El problema era que el proyecto estaba en un directorio llamado "wollok". 

El LSP está tomando como path de los archivos todo el path desde el inicio del sistema de archivos. Entonces si tenés algún nombre choto en algún momento desde la raíz hasta tu proyecto todo se chotea.

Habría que hacer lo mismo que hacemos en CLI, que es tomar un directorio base y construir todo relativo a ese, de esa forma no importa dónde esté tu proyecto, solo importa lo que tengas adentro.
